src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -330,15 +338,25 @@ private async Task RenegotiateAsync(CancellationToken cancellationToken)
                 {
                     if (status.ErrorCode == SecurityStatusPalErrorCode.NoRenegotiation)
                     {
-                        // peer does not want to renegotiate. That should keep session usable.
+                        // Peer does not want to renegotiate. That should keep session usable.
                         return;
                     }
 
                     throw SslStreamPal.GetException(status);
                 }
 
-                // Issue empty read to get renegotiation going.
-                await ReadAsyncInternal(adapter, Memory<byte>.Empty, renegotiation: true).ConfigureAwait(false);
+                _handshakeBuffer = new ArrayBuffer(InitialHandshakeBufferSize);
+                ProtocolToken message = null!;
```

Comment:
This seems similar to the existing logic in ForceAuthenticationAsync. Can we share the logic so it isn't duplicated?

I also wonder about some cases that ForceAuthenticationAsync is handling which aren't handled here, like transferring any additional buffered read data to the _internalBuffer.

Reply:
I added the missing handling of _internalBuffer. I agree about the similarity but I would like to let re-factoring for and consolidation for follow up  work so we can get the base functionality in.